File: internal/notify/model/order.go
package model

import (
	"fmt"
	"time"
)

// OrderStatus represents the state of a purchase order.
type OrderStatus string

const (
	OrderCreated        OrderStatus = "created"
	OrderPaid           OrderStatus = "paid"
	OrderConfirmed      OrderStatus = "confirmed"
	OrderShipped        OrderStatus = "shipped"
	OrderDelivered      OrderStatus = "delivered"
	OrderCancelled      OrderStatus = "cancelled"
	OrderDeliveryFailed OrderStatus = "delivery_failed"
)

// OrderType describes the demo purchase form. The value intentionally stays
// lightweight because GoIM delivery is the focus of this service.
type OrderType string

const (
	OrderTypeNormal     OrderType = "normal"
	OrderTypePresale    OrderType = "presale"
	OrderTypeUrgent     OrderType = "urgent"
	OrderTypeEnterprise OrderType = "enterprise"
	OrderTypeAfterSale  OrderType = "after_sale"
	OrderTypeVirtual    OrderType = "virtual"
)

// OrderImportance controls notification priority, TTL, and ACK policy.
type OrderImportance string

const (
	OrderImportanceNormal   OrderImportance = "normal"
	OrderImportanceHigh     OrderImportance = "high"
	OrderImportanceUrgent   OrderImportance = "urgent"
	OrderImportanceCritical OrderImportance = "critical"
)

// Order represents a demo purchase order.
type Order struct {
	OrderID               string          `json:"order_id"`
	UserID                string          `json:"user_id"`
	MerchantID            string          `json:"merchant_id,omitempty"`
	MerchantUID           int64           `json:"merchant_uid,omitempty"`
	MerchantName          string          `json:"merchant_name,omitempty"`
	Status                OrderStatus     `json:"status"`
	OrderType             OrderType       `json:"order_type,omitempty"`
	Importance            OrderImportance `json:"importance,omitempty"`
	BuyerNote             string          `json:"buyer_note,omitempty"`
	FulfillmentMode       string          `json:"fulfillment_mode,omitempty"`
	SupportRoomID         string          `json:"support_room_id,omitempty"`
	PrivateConversationID string          `json:"private_conversation_id,omitempty"`
	Items                 []OrderItem     `json:"items"`
	Total                 float64         `json:"total"`
	CreatedAt             time.Time       `json:"created_at"`
	UpdatedAt             time.Time       `json:"updated_at"`
}

// OrderItem represents a line item in an order.
type OrderItem struct {
	ProductID   string  `json:"product_id,omitempty"`
	SKUID       string  `json:"sku_id,omitempty"`
	ProductName string  `json:"product_name"`
	Quantity    int     `json:"quantity"`
	Price       float64 `json:"price"`
	ImageURL    string  `json:"image_url,omitempty"`
}

// OrderStatusEvent is an append-only record for an order state change.
type OrderStatusEvent struct {
	EventID        string            `json:"event_id"`
	OrderID        string            `json:"order_id"`
	FromStatus     OrderStatus       `json:"from_status"`
	ToStatus       OrderStatus       `json:"to_status"`
	Extra          map[string]string `json:"extra,omitempty"`
	IdempotencyKey string            `json:"idempotency_key,omitempty"`
	CreatedAt      time.Time         `json:"created_at"`
}

// orderTransitions defines valid order state transitions.
var orderTransitions = map[OrderStatus][]OrderStatus{
	OrderCreated:        {OrderConfirmed, OrderCancelled},
	OrderPaid:           {OrderConfirmed, OrderCancelled},
	OrderConfirmed:      {OrderShipped, OrderCancelled},
	OrderShipped:        {OrderDelivered, OrderDeliveryFailed},
	OrderDelivered:      {},
	OrderCancelled:      {},
	OrderDeliveryFailed: {},
}

// TransitionError describes why an order status transition is invalid.
type TransitionError struct {
	From      OrderStatus   `json:"from"`
	To        OrderStatus   `json:"to"`
	Allowed   []OrderStatus `json:"allowed"`
	Reason    string        `json:"reason"`
	Timestamp time.Time     `json:"timestamp"`
}

func (e *TransitionError) Error() string {
	return fmt.Sprintf("invalid transition from %s to %s: allowed %v", e.From, e.To, e.Allowed)
}

// ValidTransition checks if a status transition is valid.
// Returns nil if valid, or a TransitionError if invalid.
func ValidTransition(from, to OrderStatus) error {
	allowed := orderTransitions[from]
	for _, valid := range allowed {
		if valid == to {
			return nil
		}
	}
	return &TransitionError{
		From:      from,
		To:        to,
		Allowed:   allowed,
		Reason:    fmt.Sprintf("status %s cannot transition to %s", from, to),
		Timestamp: time.Now(),
	}
}

File: web/dashboard/src/types/order.ts
export type OrderStatus =
  | 'created'
  | 'paid'
  | 'confirmed'
  | 'shipped'
  | 'delivered'
  | 'cancelled'
  | 'delivery_failed'

export type OrderType =
  | 'normal'
  | 'presale'
  | 'urgent'
  | 'enterprise'
  | 'after_sale'
  | 'virtual'

export type OrderImportance = 'normal' | 'high' | 'urgent' | 'critical'

export interface Merchant {
  merchant_id: string
  merchant_uid: number
  name: string
  description?: string
  logo_url?: string
  group_room_id?: string
  group_name?: string
  created_at: string
  updated_at: string
}

export interface Product {
  product_id: string
  merchant_id: string
  sku_id?: string
  name: string
  description?: string
  price: number
  image_url?: string
  fulfillment_mode?: string
  created_at: string
  updated_at: string
}

export interface MerchantGroup {
  group_id: string
  merchant_id: string
  merchant_uid: number
  room_id: string
  name: string
  description?: string
  member_count: number
  created_at: string
  updated_at: string
}

export interface OrderItem {
  product_id?: string
  sku_id?: string
  product_name: string
  quantity: number
  price: number
  image_url?: string
}

export interface Order {
  order_id: string
  user_id: string
  merchant_id?: string
  merchant_uid?: number
  merchant_name?: string
  status: OrderStatus
  order_type?: OrderType
  importance?: OrderImportance
  buyer_note?: string
  fulfillment_mode?: string
  support_room_id?: string
  private_conversation_id?: string
  items: OrderItem[]
  total: number
  created_at: string
  updated_at: string
}

export interface PurchaseOrderItemInput {
  product_id: string
  sku_id?: string
  quantity: number
}

export interface PurchaseOrderInput {
  user_id: string
  merchant_id: string
  merchant_uid?: number
  order_type?: OrderType
  importance?: OrderImportance
  buyer_note?: string
  fulfillment_mode?: string
  items: PurchaseOrderItemInput[]
}

export const ORDER_STATUS_LABELS: Record<OrderStatus, string> = {
  created: '已下单',
  paid: '历史兼容状态',
  confirmed: '商家已确认',
  shipped: '履约已发出',
  delivered: '订单已送达',
  cancelled: '订单已取消',
  delivery_failed: '履约异常',
}

export const ORDER_TYPE_LABELS: Record<OrderType, string> = {
  normal: '普通商品',
  presale: '预售',
  urgent: '加急',
  enterprise: '企业采购',
  after_sale: '售后服务',
  virtual: '虚拟商品',
}

export const IMPORTANCE_LABELS: Record<OrderImportance, string> = {
  normal: '普通',
  high: '重要',
  urgent: '紧急',
  critical: '关键',
}

export const ORDER_STATUS_SEQUENCE: OrderStatus[] = [
  'created',
  'confirmed',
  'shipped',
  'delivered',
]

export function getValidTransitions(current: OrderStatus): OrderStatus[] {
  const transitions: Record<OrderStatus, OrderStatus[]> = {
    created: ['confirmed', 'cancelled'],
    paid: ['confirmed', 'cancelled'],
    confirmed: ['shipped', 'cancelled'],
    shipped: ['delivered', 'delivery_failed'],
    delivered: [],
    cancelled: [],
    delivery_failed: [],
  }
  return transitions[current] || []
}
